napari/config.py

```diff
@@ -0,0 +1,877 @@
+    lock : , optional
+        [description], by default config_lock
+
+    Returns
+    -------
+    synced : bool
+        Whether a sync occurred.
+    """
+    # resolve destination file
+    if destination is None:
+        destination = PATH
+    if os.path.isdir(destination):
+        destination = os.path.join(destination, '_session.yaml')
+    assert os.fspath(destination).endswith(
+        ".yaml"
+    ), "Only YAML is currently supported"
+
+    # check whether the config dict has changed since the last sync (i.e. has
+    # the key "_dirty") or whether the file on disk has been modified since
+    # last sync (i.e. has modified date greater than the key "_last_synced")
+    config_is_dirty = config.pop('_dirty', None)
+    dest_is_dirty = False
+    if os.path.exists(destination):
+        last_modified = os.path.getmtime(destination)
+        # if it has never been synced before, set _last_synced to 0 (= "never")
+        last_synced = config.get('_last_synced', 0)
+        if last_modified > last_synced:
+            dest_is_dirty = True
+    else:
+        config_is_dirty = True
+
+    # if nothing has happened, return
+    if not (config_is_dirty or dest_is_dirty):
+        return False
+
+    with lock:  # aquire file lock on yaml file
+        if dest_is_dirty:  # yaml file has changed since last sync
+            with open(destination, 'r') as f:
+                yaml_config = yaml.safe_load(f) or {}
+        else:
+            yaml_config = {}
+        if config_is_dirty:  # config has changed since last sync
+            # merge prefers values in the later arguments, over the earlier
+            # so if new keys have shown up on disk, we use them, otherwise
+            # we overwrite the disk values with the current config
+            priority = "old" if prefer_config else "new"
+            update(config, yaml_config, priority=priority)
+        elif yaml_config:
+            # if config is not dirty, but the outputfile IS dirty, we overwrite
+            # the the config with the new values from disk.
+            config.clear()
+            update(config, yaml_config)
+
+        # write the merged config to disk
+        with open(destination, 'w') as f:
+            out = config.copy()
+            out.pop('_last_synced', None)
+            if out:
+                try:
+                    yaml.dump(out, f, Dumper=ConfigDumper)
+                except yaml.YAMLError as exc:
+                    msg = f"Failed to write session config to disk: {exc}"
+                    raise type(exc)(msg)
+
+            else:  # instead of writing "{}" to file, write "# empty"
+                f.write("# empty")
+
+    # update the config with the time of last modification
+    config['_last_synced'] = os.path.getmtime(destination)
+    return True
+
+
+refresh()
+
+# read in the default settings from this directory
+fn = os.path.join(os.path.dirname(__file__), "napari.yaml")
```

Comment:
I kinda dislike having this file in `napari/napari.yaml`. How about making a directory `napari/config/{config.py,napari.yaml,...}`, and in `napari/__init__.py` doing `from .config import config`?

Reply:
I moved everything into a folder, but ran into an issue with the constant `napari.config.PATH`, which should be modifiable at that location.  If I put the whole config file in `napari/config/__init__.py` it's fine, but if I put it all in `config.py` then if a user changes `napari.config.PATH` (rather than `napari.config.config.PATH`) it breaks.  how much do you hate putting it all in `napari/config/__init__.py` vs jumping through hoops to avoid circular imports?

Reply:
I'll let @jni weigh in on this more, but personally I don't love having everything in `napari/config/__init__.py`. Maybe I don't understand how the whole config works, but can we not break this out into more files like we do elsewhere? It's ok if some stuff still lives in the `__init__.py` if that really does make things simpler, but does everything have too?

For example could we have a `config_utils.py` with some of the `dict` -> `dict` functions like `update` in there or a `str->str` function `canonical_name`.

Could we also have the classes be their own files like `_set`  and `SyncStatus`?

Reply:
> can we not break this out into more files like we do elsewhere

probably yes, but it's made trickier by the fact that this config module is very much a *stateful* module, and many of the functions in the module take, as their default values, variables that are declared within the module (for instance: `def get(key, config=config)` is using the module level `config` variable as its default).
Combine that with the desire to keep the user-facing API simple (by importing everything in the `__init__.py`) and it makes it harder to _both_ avoid circular imports _and_ keep the API nice while preventing scenarios like this:
```python
# config/__init__.py
from .inner import PATH, show_path

# config/inner.py
PATH = 'original value'

def show_path(path=PATH):
    print(path)


In [1]: from napari import config                                               
In [2]: config.show_path()                                                     
original value

In [3]: config.PATH                                                             
Out[3]: 'original value'

In [4]: config.PATH = 'new value'                                               

In [5]: config.show_path()                                                     
original value
```

This is the case with `test_ensure_file_defaults_to_NAPARI_CONFIG_directory` that I mentioned in my other comment.  Now we can definitely just say "we're not going to support changing PATH that way"... but we need to consciously decide that.